src/module_3/evaluation_functions.py

```diff
@@ -0,0 +1,179 @@
+    if hasattr(model, 'predict_proba'):
+        model_probs = model.predict_proba(X_train)
+        model_probs = model_probs[:, 1]
+        model_probs_val = model.predict_proba(X_val)
+        model_probs_val = model_probs_val[:, 1]
+    else:
+        model_probs = model.decision_function(X_train)
+        model_probs_val = model.decision_function(X_val)
+        
+    if (metric == 'fscore'):
+        model_precision, model_recall, thresholds = metrics.precision_recall_curve(y_train, model_probs)  
+        performance_thresholds = [metrics.f1_score(y_val, threshold_classification(model_probs_val, t)) * 100.0 for t in thresholds]
+        best_threshold_index = np.argmax(performance_thresholds)
+    elif (metric == 'auc'):
+        model_fpr, model_tpr, thresholds = metrics.roc_curve(y_train, model_probs)
+        performance_thresholds = [metrics.roc_auc_score(y_val, threshold_classification(model_probs_val, t)) * 100.0 for t in thresholds]
+        best_threshold_index = np.argmax(performance_thresholds)
+    elif (metric == 'fpr'):
+        model_fpr, model_tpr, thresholds = metrics.roc_curve(y_train, model_probs)
+        performance_thresholds = np.array([100 - metrics.recall_score(y_val, threshold_classification(model_probs_val, t), pos_label=1) * 100.0 for t in thresholds])
+        valid_thresholds = np.where(performance_thresholds <= 6.0)
+        best_threshold_index = valid_thresholds[0][0]
+    
+    best_threshold = thresholds[best_threshold_index]
+    print('Best threshold={:.3f}, {} on validation={:.2f}'.format(best_threshold, metric, performance_thresholds[best_threshold_index]))
+    return best_threshold
+
+def evaluate_best_threshold_test(model, threshold, X_test, y_test, metric='fscore'):
+    if hasattr(model, 'predict_proba'):
+        model_probs_test = model.predict_proba(X_test)
+        model_probs_test = model_probs_test[:, 1]
+    else:
+        model_probs_test = model.decision_function(X_test)
+        
+    if (metric == 'fscore'):
+        print('Fscore of the best model on test after optimizing threshold {:.2f}'.format(metrics.f1_score(y_test, threshold_classification(model_probs_test, threshold)) * 100.0))
+    elif (metric == 'recall_mean'):
+        print('Mean recall by classes of the best model on test after optimizing threshold {:.2f}'.format(metrics.roc_auc_score(y_test, threshold_classification(model_probs_test, threshold)) * 100.0))
+    elif (metric == 'fpr'):
+        print('FPR of the best model on test after optimizing threshold {:.2f}'.format(100 - metrics.recall_score(y_test, threshold_classification(model_probs_test, threshold), pos_label=1) * 100.0))
+    predictions_test_best_threshold = threshold_classification(model_probs_test, threshold)
+    print(metrics.confusion_matrix(y_test, predictions_test_best_threshold, labels=[1, 0]))
+
+def plot_validation_curve(train_scores, test_scores, param_range, ax=None, param='Param', score='Score', scale=None):
```

Comment:
It's not obvious at all what you are doing here. If you are plotting something that is somewhat complex you should explain what it is with docstrings.

arguments should have type hints too to make it easier to understand.